Added `int.signum`, `float.signum`, `float.is-nan`, and `float.is-infinite`

Dherse opened the pull request:
Closes #3113 

I went for the first approach suggested @PgBiel of specializing `calc.sign` to work differently whether it's a float or an integer. This can of course be trivially changed.

PgBiel commented:
I guess one problem of always returning integer, instead of returning a float when a float is given, is that we can't return `NaN` when `NaN` is given...

Maybe we could follow Rust fully and return a float when given a float (including `NaN` when given `NaN`)?

Although `NaN` is honestly a huge pain to work with in Typst, so I'm not sure if such behavior would be desired... (though, to get here you'd need a `NaN` in the first place anyway :p)

laurmaedje commented:
Based on the discussion in https://github.com/typst/typst/issues/3113, I think it'd be preferrable to return int for int and float for float. Moreover, I wonder why the name changed from `signum` in the issue to `sign` here. Which one is more common in other programming languages? Maybe there is some relevant previous discussion about naming in Rust's issues.

Dherse commented:
> Based on the discussion in #3113, I think it'd be preferrable to return int for int and float for float. Moreover, I wonder why the name changed from `signum` in the issue to `sign` here. Which one is more common in other programming languages? Maybe there is some relevant previous discussion about naming in Rust's issues.

I have renamed it to `signum`, although I generally prefer `sign` but I'm not a attached to either one specifically. I have also adjusted the behavior to match exactly rust's. Because of this I also added a `is-nan` method to `float` to let the user check for `nan`s, maybe that could be replaced by returning `0.0` for a nan, idk

istudyatuni commented on crates/typst/src/foundations/float.rs:
````diff
@@ -47,6 +47,21 @@ impl f64 {
+    /// #float.is-nan(1) \
+    /// #float.is-nan(calc.nan)
+    /// ```
+    #[func]
+    pub fn is_nan(self) -> bool {
+        self.is_nan()
````
How does this work? `f64` type is redefined, so how is it possible to call `f64::is_nan`? Is it because of macro?

Dherse replied:
It works using the macros

laurmaedje replied:
The `type f64` is erased by the macro, it's just the to communicate the type to the macro. For primitive types like this the `#[scope]` macro generates an extension trait under the hood to put the functions into. Since is_nan is defined as an inherent function on f64, `self.is_nan()` dispatches to that one instead of the one new one in the extension trait. It took me a bit to figure that out what's happening though, so perhaps it could be made clearer that this isn't infinitely recursive by writing `f64::is_nan(self)` instead.

istudyatuni commented on crates/typst/src/foundations/calc.rs:
```diff
@@ -961,6 +994,18 @@ cast! {
     v: f64 => Self::Float(v),
 }
 
+impl From<i64> for Num {
+    fn from(value: i64) -> Self {
+        Self::Int(value)
+    }
+}
+
+impl From<f64> for Num {
+    fn from(value: f64) -> Self {
+        Self::Float(value)
+    }
+}
```
Why is this need to be defined if this is already defined above in `cast!`?

laurmaedje replied:
cast only implements FromValue and IntoValue. still, I think rather than having these impls (although they don't hurt), it would be nice to have an apply1 function on Num which we could use to shorten signum.

laurmaedje commented:
> I have renamed it to signum, although I generally prefer sign but I'm not a attached to either one specifically.

I'm also not attached to either one. I just wanted to raise attention to the fact that it changed.

Dherse commented:
@PgBiel also requested on Discord whether `float.is-infinite()` could be added so I included it in this PR, it feels like it fits, it's more improvements of floats than anything else.

laurmaedje commented on crates/typst/src/foundations/float.rs:
```diff
@@ -47,6 +47,36 @@ impl f64 {
+
+    /// Checks if a float is not a number.
+    ///
+    /// In IEEE 754, not all bit patterns are valid floats. This function
```
Is this phrasing technically correct? Isn't NaN also a valid float?

laurmaedje commented on crates/typst/src/foundations/calc.rs:
```diff
@@ -910,6 +911,38 @@ pub fn quo(
+    /// The number whose sign to calculate.
+    value: Num,
+) -> Num {
+    match value {
+        Num::Int(n) => n.signum().into(),
+        Num::Float(n) => n.signum().into(),
+    }
```
I think we should add apply1 and use it here.

laurmaedje commented:
I wonder whether we should add signum to calc for the time being or already start here with opting for methods...

PgBiel commented:
> I wonder whether we should add signum to calc for the time being or already start here with opting for methods...

I was thinking about this too... Maybe we can just straight up avoid the inevitable breaking changes.

Although it might seem inconsistent to users... but we can just explain what we're doing in `calc`'s docs.

laurmaedje commented:
Thanks!
